Description: Sync chosen background filename to accountsservice
 used in lightdm by unity-greeter
Bug: https://launchpad.net/bugs/1128492
Bug: https://launchpad.net/bugs/928553
---
 nautilus-desktop/nautilus-desktop-background.c |   81 +++++++++++++++++++++++++
 1 file changed, 81 insertions(+)

--- a/nautilus-desktop/nautilus-desktop-background.c
+++ b/nautilus-desktop/nautilus-desktop-background.c
@@ -75,6 +75,85 @@ struct NautilusDesktopBackgroundDetails
 };
 
 
+static void
+set_accountsservice_background (const gchar *background)
+{
+  GDBusProxy *proxy = NULL;
+  GDBusProxy *user = NULL;
+  GVariant *variant = NULL;
+  GError *error = NULL;
+  gchar *object_path = NULL;
+
+  proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SYSTEM,
+                                         G_DBUS_PROXY_FLAGS_NONE,
+                                         NULL,
+                                         "org.freedesktop.Accounts",
+                                         "/org/freedesktop/Accounts",
+                                         "org.freedesktop.Accounts",
+                                         NULL,
+                                         &error);
+
+  if (proxy == NULL) {
+    g_warning ("Failed to contact accounts service: %s", error->message);
+    g_error_free (error);
+    return;
+  }
+
+  variant = g_dbus_proxy_call_sync (proxy,
+                                    "FindUserByName",
+                                    g_variant_new ("(s)", g_get_user_name ()),
+                                    G_DBUS_CALL_FLAGS_NONE,
+                                    -1,
+                                    NULL,
+                                    &error);
+
+  if (variant == NULL) {
+    g_warning ("Could not contact accounts service to look up '%s': %s",
+               g_get_user_name (), error->message);
+    g_error_free (error);
+    goto bail;
+  }
+
+  g_variant_get (variant, "(o)", &object_path);
+  user = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SYSTEM,
+                                        G_DBUS_PROXY_FLAGS_NONE,
+                                        NULL,
+                                        "org.freedesktop.Accounts",
+                                        object_path,
+                                        "org.freedesktop.Accounts.User",
+                                        NULL,
+                                        &error);
+  g_free (object_path);
+
+  if (user == NULL) {
+    g_warning ("Could not create proxy for user '%s': %s",
+               g_variant_get_string (variant, NULL), error->message);
+    g_error_free (error);
+    goto bail;
+  }
+  g_variant_unref (variant);
+
+  variant = g_dbus_proxy_call_sync (user,
+                                    "SetBackgroundFile",
+                                    g_variant_new ("(s)", background ? background : ""),
+                                    G_DBUS_CALL_FLAGS_NONE,
+                                    -1,
+                                    NULL,
+                                    &error);
+
+  if (variant == NULL) {
+    g_warning ("Failed to set the background '%s': %s", background, error->message);
+    g_error_free (error);
+    goto bail;
+  }
+
+bail:
+  if (proxy != NULL)
+    g_object_unref (proxy);
+  if (variant != NULL)
+    g_variant_unref (variant);
+}
+
 static gboolean
 background_settings_change_event_cb (GSettings *settings,
                                      gpointer   keys,
@@ -460,6 +539,8 @@ background_change_event_idle_cb (Nautilu
     gnome_bg_load_from_preferences (self->details->bg,
                                     gnome_background_preferences);
 
+    set_accountsservice_background (gnome_bg_get_filename (self->details->bg));
+
     g_object_unref (self);
 
     return FALSE;
